File: api/src/middleware/metrics.ts
import { FastifyInstance } from 'fastify';
import { collectDefaultMetrics, Counter, Histogram, Gauge, register } from 'prom-client';

collectDefaultMetrics({ prefix: 'obs_' });

export const httpRequestCounter = new Counter({
  name: 'obs_http_requests_total',
  help: 'Total HTTP requests',
  labelNames: ['method', 'route', 'status'],
});

export const httpRequestDuration = new Histogram({
  name: 'obs_http_request_duration_ms',
  help: 'HTTP request duration in ms',
  labelNames: ['method', 'route'],
  buckets: [5, 10, 25, 50, 100, 250, 500, 1000, 2500, 5000],
});

export const activeSessionsGauge = new Gauge({
  name: 'obs_active_sessions',
  help: 'Currently active recording sessions',
});

export const kafkaMessagesCounter = new Counter({
  name: 'obs_kafka_messages_total',
  help: 'Kafka messages produced/consumed',
  labelNames: ['topic', 'direction'],
});

export const indexedDocumentsCounter = new Counter({
  name: 'obs_opensearch_indexed_total',
  help: 'Documents indexed to OpenSearch',
  labelNames: ['index'],
});

export function registerMetricsRoute(app: FastifyInstance): void {
  app.get('/api/metrics', async (_, reply) => {
    reply.header('Content-Type', register.contentType);
    return register.metrics();
  });
}

File: api-rust/src/metrics.rs
use lazy_static::lazy_static;
use prometheus::{opts, register_counter_vec, register_histogram_vec, CounterVec, HistogramVec};

lazy_static! {
    pub static ref HTTP_REQUESTS_TOTAL: CounterVec = register_counter_vec!(
        opts!("obs_http_requests_total", "Total HTTP requests"),
        &["method", "route", "status"]
    )
    .unwrap();
    pub static ref HTTP_REQUEST_DURATION_MS: HistogramVec = register_histogram_vec!(
        "obs_http_request_duration_ms",
        "HTTP request duration in ms",
        &["method", "route"],
        vec![5.0, 10.0, 25.0, 50.0, 100.0, 250.0, 500.0, 1000.0, 2500.0, 5000.0]
    )
    .unwrap();
    pub static ref ACTIVE_SESSIONS: prometheus::Gauge = prometheus::register_gauge!(
        "obs_active_sessions",
        "Number of active monitoring sessions"
    )
    .unwrap();
    pub static ref KAFKA_MESSAGES_TOTAL: CounterVec = register_counter_vec!(
        opts!("obs_kafka_messages_total", "Total Kafka messages processed"),
        &["topic", "direction"]
    )
    .unwrap();
    pub static ref OPENSEARCH_INDEXED_TOTAL: CounterVec = register_counter_vec!(
        opts!("obs_opensearch_indexed_total", "Total OpenSearch documents indexed"),
        &["index"]
    )
    .unwrap();
}

#[derive(Clone)]
pub struct Metrics;

impl Metrics {
    pub fn new() -> Self {
        Metrics
    }
}

